addons/account/models/chart_template.py

```diff
@@ -948,14 +948,132 @@ def name_get(self):
             res.append((record.id, name))
         return res
 
+    @api.model
+    def try_instantiating_foreign_taxes(self, country, companies):
+        """ This function is called in multivat setup, when a company needs to submit a
+        tax report in a foreign country.
+
+        It searches for tax templates in the provided countries and instantiates the
+        ones it find in each of the provided companies.
+
+        Tax accounts are not kept from the templates (this wouldn't make sense,
+        as they don't belong to the same CoA as the one installed on the company).
+        Instead, we search existing tax accounts for approximately equivalent accounts
+        and use their prefix to create new accounts. Doing this gives a roughly correct suggestion
+        that then needs to be reviewed by the user to ensure its consistency.
+        It is intended as a shortcut to avoid hours of encoding, not as an out-of-the-box, always
+        correct solution.
+        """
+        def create_foreign_tax_account(existing_account, additional_label):
+            new_code = self.env['account.account']._search_new_account_code(existing_account.company_id, len(existing_account.code), existing_account.code[:-2])
+            return self.env['account.account'].create({
+                'name': '%s - %s' % (existing_account.name, additional_label),
+                'code': new_code,
+                'user_type_id': existing_account.user_type_id.id,
+                'company_id': existing_account.company_id.id,
+            })
+
+        def get_existing_tax_account(foreign_tax_rep_line, force_tax=None):
+            company = foreign_tax_rep_line.company_id
+            sign_comparator = '<' if foreign_tax_rep_line.factor_percent < 0 else '>'
+
+            search_domain = [
+                ('account_id', '!=', False),
+                ('factor_percent', sign_comparator, 0),
+                ('company_id', '=', company.id),
+                '|',
+                '&', ('invoice_tax_id.type_tax_use', '=', tax_rep_line.invoice_tax_id.type_tax_use),
+                     ('invoice_tax_id.country_id', '=', company.account_fiscal_country_id.id),
+                '&', ('refund_tax_id.type_tax_use', '=', tax_rep_line.refund_tax_id.type_tax_use),
+                     ('refund_tax_id.country_id', '=', company.account_fiscal_country_id.id),
+            ]
+
+            if force_tax:
+                search_domain += [
+                    '|', ('invoice_tax_id.id', 'in', force_tax.ids),
+                    ('refund_tax_id.id', 'in', force_tax.ids),
+                ]
+
+            return self.env['account.tax.repartition.line'].search(search_domain, limit=1).account_id
+
+
+        taxes_in_country = self.env['account.tax'].search([
+            ('country_id', '=', country.id),
+            ('company_id', 'in', companies.ids)
+        ])
+
+        if taxes_in_country:
+            return
```

Comment:
shouldn't be done case by case depending the company? In practice, do we really want this method called on multiple companies?

Reply:
In the original version of the spec, we did. But not anymore indeed. I replaced companies by company and removed the loop.